> ### **Assignment 2 - Numpy Array Operations** 
>
> This assignment is part of the course ["Data Analysis with Python: Zero to Pandas"](http://zerotopandas.com). The objective of this assignment is to develop a solid understanding of Numpy array operations. In this assignment you will:
> 
> 1. Pick 5 interesting Numpy array functions by going through the documentation: https://numpy.org/doc/stable/reference/routines.html 
> 2. Run and modify this Jupyter notebook to illustrate their usage (some explanation and 3 examples for each function). Use your imagination to come up with interesting and unique examples.
> 3. Upload this notebook to your Jovian profile using `jovian.commit` and make a submission here: https://jovian.ml/learn/data-analysis-with-python-zero-to-pandas/assignment/assignment-2-numpy-array-operations
> 4. (Optional) Share your notebook online (on Twitter, LinkedIn, Facebook) and on the community forum thread: https://jovian.ml/forum/t/assignment-2-numpy-array-operations-share-your-work/10575 . 
> 5. (Optional) Check out the notebooks [shared by other participants](https://jovian.ml/forum/t/assignment-2-numpy-array-operations-share-your-work/10575) and give feedback & appreciation.
>
> The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.
>
> Try to give your notebook a catchy title & subtitle e.g. "All about Numpy array operations", "5 Numpy functions you didn't know you needed", "A beginner's guide to broadcasting in Numpy", "Interesting ways to create Numpy arrays", "Trigonometic functions in Numpy", "How to use Python for Linear Algebra" etc.
>
> **NOTE**: Remove this block of explanation text before submitting or sharing your notebook online - to make it more presentable.


# Title Here


### Subtitle Here

Write a short introduction about Numpy and list the chosen functions. 

- function 1
- function 2
- function 3
- function 4
- function 5

The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.

In [5]:
!pip install jovian --upgrade -q

In [6]:
import jovian

In [7]:
jovian.commit(project='numpy-array-operations')

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "arunsalamander/numpy-array-operations" on https://jovian.ai/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/arunsalamander/numpy-array-operations


'https://jovian.ai/arunsalamander/numpy-array-operations'

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [8]:
import numpy as np

In [9]:
# List of functions explained 
function1 = np.diag  
function2 = np.flip
function3 = np.sort
function4 = np.bitwise_and
function5 = np.packbits

## Function 1 - np.concatenate (change this)

Add some explanation about the function in your own words

In [6]:
# Example 1 - working : Extracts diagonal from a matrix
arr= np.array([[1,2,3],
               [3,5,6],
               [5,2,1]])

#extracts the diagonal value
print(np.diag(arr))


[1 5 1]


#### Based on k value, we can extract other diagonals

In [7]:
# Example 2 - working
arr=np.array([[1,2,4],
              [2,6,7],
              [5,6,9]])

#k=-1

print('Diagonal below main diagonal:',np.diag(arr,k=-1))

#k=1

print('Diagonal above main diagonal:',np.diag(arr,k=1))

Diagonal below main diagonal: [2 6]
Diagonal above main diagonal: [2 7]


Explanation about example

In [8]:
# Example 3 - breaking (to illustrate when it breaks)
arr=([[1,2,3],
     [4,6,7],
     [8,9,4]])

np.diag(arr,k=1.0)

#value of k parameter should be integer

TypeError: integer argument expected, got float

The code breaks as the function was excepting a int as the argument but it got float.

This functin can be used for extracting diagonal elements and creating diagonal arrays such as Identity matrix

In [ ]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..


## Function 2 - numpy.flip

This function reverses the order of elements in an array along the given axis y preserving the shape of the array but the elements are reordered

In [9]:
# Example 1 - flipping entire array
arr= np.array([[2,3,5],
              [6,5,7],
              [7,8,4]])
np.flip(arr)

array([[4, 8, 7],
       [7, 5, 6],
       [5, 3, 2]])

numpy.flip reverses the array also the order of the element changes

In [11]:
# Example 2 - working : flipping array in a particular direction
arr= np.array([[11,45,23],
              [34,56,74],
              [5,6,7]])

np.flip(arr,1)

array([[23, 45, 11],
       [74, 56, 34],
       [ 7,  6,  5]])

Here the elements are flipped in row or in the left-right direction. we can also flip the elements in up-down direction by giving the 2nd parameter, i.e, 1 or 0

In [14]:
# Example 3 - breaking
arr= np.array([[11,45,23],
              [34,56,74],
              [5,6,7]])

np.flip(arr,2)

AxisError: axis 2 is out of bounds for array of dimension 2

The code breaks because 2D matrix has only 2 axis 1 axis and 0 axis, where as third axis i.e, z- axis doesnot fit in the above example, the value 2 will be considered for a 3D matrix


Useful to flip the matrices

In [15]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "arunsalamander/numpy-array-operations" on https://jovian.ai/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/arunsalamander/numpy-array-operations


'https://jovian.ai/arunsalamander/numpy-array-operations'

## Function 3 - numpy.sort

This function returns a copy of the sorted array

In [16]:
# Example 1 - working

arr= np.array([[1,5,2],
              [4,5,7],
              [5,6,7]])

np.sort(arr)

array([[1, 2, 5],
       [4, 5, 7],
       [5, 6, 7]])

numpy.sort , sorts the matrix elements in ascending order by default

In [18]:
# Example 2 - sorting in reverse order
arr= np.array([[11,45,23],
              [34,56,74],
              [5,6,7]])
np.sort(arr, kind='mergesort',axis=-1)[:,::-1]

array([[45, 23, 11],
       [74, 56, 34],
       [ 7,  6,  5]])

Here, we sorted the array using merge sort and reversed the array using the list slicing 

In [19]:
# Example 3 - breaking (to illustrate when it breaks)
arr= np.array([[1,'hi'],
             ['bye',3.0,9],
             ['yo',8,6]])
np.sort(arr)

<ipython-input-19-9f309069c27f>:2: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  arr= np.array([[1,'hi'],


TypeError: '<' not supported between instances of 'str' and 'int'

Here, we have array with different number of elements in each list, hence it throws an error, when the number of elements are same we will get a sorted array.

This a really useful in sorting arrays

In [ ]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..


## Function 4- numpy.bitwise_and

This function returns the values after performing a bitwise and operation 
T^T=T

In [10]:
# Example 1 - working

arr1= ([[1,3,5],
      [3,4,3],
      [0,6,4]])

arr2= ([[1,1,1],
       [1,1,1],
       [1,1,1]])

np.bitwise_and(arr1,1)

array([[1, 1, 1],
       [1, 0, 1],
       [0, 0, 0]])

Here the function first converts the arr1 elements to their binary representation and then performs AND operation

In [12]:
# Example 2 - working

arr1=([[1,0,3],
     [4,5,6],
     [0,8,9]])

np.bitwise_and(arr1,1)

array([[1, 0, 1],
       [0, 1, 0],
       [0, 0, 1]])

we can also achieve the same type of output by simply providing the number by which we want to AND to the function arguments

In [13]:
# Example 3 - breaking (to illustrate when it breaks)

arr1= ([[1,0,3],
       [4,5,6],
       [0,8,9]])

arr2=([[1,1,'hello'],
      [1,1,1],
      [1,1,1]])

np.bitwise_and(arr1,arr2)

TypeError: ufunc 'bitwise_and' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

The code breaks as AND operation is only performed when there are integer that can be converted to binary number, the code will run again if we give and number other than float to the arr2

This function bitwise performs bitwise AND operation and returns the result

In [14]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "arunsalamander/numpy-array-operations" on https://jovian.ai/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/arunsalamander/numpy-array-operations


'https://jovian.ai/arunsalamander/numpy-array-operations'

## Function 5 - numpy.packbits

This function returns a binary-valued array into bits in a uint8 array. The result is padded to full bytes by inserting zero bits at the end

In [17]:
# Example 1 - working

arr1= np.array([[[1,0,1],
               [0,1,0]],
               [[1,1,0],
               [0,0,1]]])

np.packbits(arr1,axis=0)

array([[[192,  64, 128],
        [  0, 128,  64]]], dtype=uint8)

Here, we have a 3D array so, we would have 3 axis for the 0th(x) axis the function takes the first element at 'a[0][0][0]' -- '1' and second element 'a[1][0][0] -- '1' and appends six 0's since the function works with 9 bit binary packing and hence we get 192 which is -- 11000000 in binary

In [20]:
# Example 2 - working

arr1= np.array([[[1,0,1],
                 [0,1,0]],
                [[1,1,0],
                 [0,0,1]]])

np.packbits(arr1,axis=1)

array([[[128,  64, 128]],

       [[128, 128,  64]]], dtype=uint8)

Using the same array in above example, we give axis1 (y) and now the function takes the first element at 'a[0][0][0]' --> '1' and second element at 'a[0][0][0]' --> '0' and appends six 0's and returns the packed value as 128--> 10000000 in binary

In [21]:
# Example 3 - breaking (to illustrate when it breaks)
arr1= np.array(([[[1,0,1,0],
                  [0,1,0]],
                 [[1,1,0],
                 [0,0,1]]]))
np.packbits(arr1,axis=1)

<ipython-input-21-84f144ef95fd>:2: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  arr1= np.array(([[[1,0,1,0],


TypeError: Expected an input array of integer or boolean data type

The function breaks as it is expecting a int array but we have a float array it can be easily solved by replacing the floating number to an integer and the matrix should always contain equal no og elemts ie the same number of elements in specific direction if taking the z(2) axis, there should be 3 numbers and for x, y 2 numbers should be present

In [23]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "arunsalamander/numpy-array-operations" on https://jovian.ai/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/arunsalamander/numpy-array-operations


'https://jovian.ai/arunsalamander/numpy-array-operations'

## Conclusion

Summarize what was covered in this notebook, and where to go next

## Reference Links
Provide links to your references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* ...

In [ ]:
jovian.commit()

<IPython.core.display.Javascript object>